[deliver] upload_screenshots_spec to test screenshot collection

liamnichols opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

The iPad Pro 12.9" (3rd Generation) has the same resolution as the 2nd and 1st Generation versions however in App Store Connect, they have different slots (and `device_type` values) for screenshots and this breaks a fundamental assumption that `deliver` has made when uploading screenshots.  

#14116 is tracking the actual issue itself and its pretty certain that `app_screenshot.rb` and `upload_screenshots.rb` will need to be updated to account for this however while I was investigating, I struggled undemanding all of the complicated things that these modules are handling so decided to try and write some tests for the existing behaviour.

### Description

These changes simply introduce an additional spec to test the collection of screenshots by `Deliver::UploadScreenshots`. The following cases are tested: 

- An empty directory (no screenshots)
- The most basic case of a screenshot
- Excluding device_type's unsupported by App Store Connect (the iPhone XR)
- Localisation
- Excluding original screenshots if _framed variants exist
  - But preserving Apple Watch screenshots without frames
- The magic appleTV directory
- iMessage screenshots

Since this module relies heavily on the file system, I copied the approach used in loader_spec.rb and used FakeFS to simulate things. Not sure if it's the best approach but it seems to have worked!

janpio commented on deliver/lib/deliver/upload_screenshots.rb:
```diff
@@ -146,5 +146,13 @@ def collect_screenshots_for_languages(path, ignore_validation)
       return screenshots
     end
+
+    def self.available_languages
```
Interesting. Why was this change necessary?

liamnichols replied:
I copied the same approach from loader.rb 

https://github.com/liamnichols/fastlane/blob/2fabb8ec9f84a19ebb00af47ac6875ca4fdb576f/deliver/lib/deliver/loader.rb#L27-L31

Since this code path will run through the tests, I didn’t want it to try and fetch this via the ITC api...

Although now you point it out, I guess I could just mock that spaceship method in the spec instead? Would it be better to do that? 🤔

giginet replied:
It can be `private`

liamnichols replied:
I think I prefer the idea of mocking it now that you mention it and since that is what I’m doing for the file size it probably makes even more sense... I’ll update the branch to do this before merging 👍🏼

liamnichols replied:
So for some reason `Spaceship::Tunes.client` is `nil` during the tests so it involves more than just stubbing the `available_languages` property on the class as I have to also initialise it in the tests and I'm not really sure how to do that 😅 

I've just reverted it back to how it was and used the `Helper.test?` check again

janpio replied:
That's fine, maybe create an issue for what you tried to achieve so we don't forget about it and can maybe take another shot later?

giginet commented on deliver/spec/upload_screenshots_spec.rb:
```diff
@@ -0,0 +1,86 @@
+
+    def add_screenshot(file)
+      FileUtils.mkdir_p(File.dirname(file))
+      File.open(file, 'w') { |f| f << 'touch' }
```
Should this be binary?

liamnichols replied:
I'm not sure.. I copied this approach from `loader_spec.rb#L14` so I hope its fine

giginet reviewed (approved):
🚀

liamnichols committed:
Introduce upload_screenshots_spec to test screenshot collection

Add some documentation explaining reasoning for UploadScreenshots.available_language

Co-Authored-By: liamnichols <liam.nichols.ln@gmail.com>

State the specific method being tested in the spec

Co-Authored-By: liamnichols <liam.nichols.ln@gmail.com>

liamnichols commented:
Whoops ☝️ I tried making the changes on an iPad without actually testing the code 🙈 ... will fix it tomorrow

liamnichols commented:
@giginet .. I tried to apply the `private` attribute but got the following error:

> private (on line 150) does not make singleton methods private. Use private_class_method or private inside a class << self block instead.
Introduce upload_screenshots_spec to test screenshot collection

I used `private_class_method` instead and it caused deployment tests to fail 🙈 
I'm not sure if there was a better way to do this but my Ruby abilities are holding me back, do you still want me to make the changes or are you happy to merge? Thanks 🙇

janpio commented on deliver/lib/deliver/upload_screenshots.rb:
```diff
@@ -146,5 +146,13 @@ def collect_screenshots_for_languages(path, ignore_validation)
       return screenshots
     end
+
+    def self.available_languages
```
```suggestion
    # helper method so Spaceship::Tunes.client.available_languages is easier to test
    def self.available_languages
```
maybe?

janpio commented on deliver/spec/upload_screenshots_spec.rb:
```diff
@@ -0,0 +1,93 @@
+require 'fakefs/spec_helpers'
+
+describe Deliver::UploadScreenshots do
+  describe "screenshot collection" do
```
```suggestion
  describe "#collect_screenshots_for_languages (screenshot collection)" do
```
Is this correct ruby convention?

janpio reviewed (approved):
Looks good to me. 

Not sure if this covers 100% of the method, but we can figure that out later - this is much better than no tests at all.

giginet reviewed (approved):
> I used private_class_method instead and it caused deployment tests to fail 🙈

I got.
You can also test for private methods using `Object#send`, but this approach is tricky and complex. you not need to make `private`.

Thank you 👍

janpio commented:
(No need to clean up your commits and force push @liamnichols btw - we always "Squash and merge" so that a PR always gets merged as one clean commit)

janpio commented:
❤️ 😍 💟 New tests!  😻 💓 🚀

liamnichols commented:
> (No need to clean up your commits and force push @liamnichols btw - we always "Squash and merge" so that a PR always gets merged as one clean commit)

This is good to know 😬 

Thanks for all the help with things!
